File: packages/server/src/commands/GitRebase.ts
import {resolve, isAbsolute, sep} from 'path';
import {existsSync} from 'fs';
import {getAgentEnv} from './SshAgentInit';
import type {IWsRequest} from '@git-yak/protocol';

function validateRepoPath(repoPath: unknown): string {
	if (typeof repoPath !== 'string' || !repoPath) {
		throw new Error('repo_path must be a non-empty string');
	}

	const resolved = isAbsolute(repoPath) ? repoPath : resolve(repoPath);

	if (!existsSync(resolved)) {
		throw new Error(`Repository path does not exist: ${resolved}`);
	}

	return resolved;
}

// The todo path becomes part of a `sequence.editor` shell command, so it must be proven to
// sit inside the repository before it is interpolated. A path that escapes the repo — or
// contains a quote that could break out of the surrounding single quotes — is refused.
function validateTodoPath(repoRoot: string, todoPath: unknown): string {
	if (typeof todoPath !== 'string' || !todoPath) {
		throw new Error('todo_path must be a non-empty string');
	}

	if (todoPath.includes("'")) {
		throw new Error('todo_path must not contain a single quote');
	}

	const resolved = resolve(repoRoot, todoPath);

	if (resolved !== repoRoot && !resolved.startsWith(repoRoot + sep)) {
		throw new Error(`Access denied: todo path outside repository: ${resolved}`);
	}

	if (!existsSync(resolved)) {
		throw new Error(`Todo file does not exist: ${resolved}`);
	}

	return resolved;
}

function buildArgs(action: string, repoRoot: string, data: IWsRequest): Array<string> {
	if (action === 'continue') {
		// core.editor=true accepts the in-progress message as-is, with no prompt.
		return ['-c', 'core.editor=true', 'rebase', '--continue'];
	}

	if (action !== 'start') {
		throw new Error(`Unknown rebase action: ${action}`);
	}

	const {upstream} = data;

	if (typeof upstream !== 'string' || !upstream) {
		throw new Error('upstream must be a non-empty string');
	}

	if (upstream.startsWith('-')) {
		throw new Error('upstream must not start with "-"');
	}

	const todo = validateTodoPath(repoRoot, data['todo_path']);

	// Our pre-written todo is copied over git's generated one via sequence.editor, and
	// core.editor is disabled so no action ever opens an interactive editor.
	return [
		'-c', 'core.editor=false',
		'-c', 'rebase.missingCommitsCheck=ignore',
		'-c', `sequence.editor=cp '${todo}'`,
		'rebase', '-i', '--autostash', '--', upstream,
	];
}

export async function run(ws: {send: (msg: string) => void}, data: IWsRequest): Promise<void> {
	const {requestId} = data;

	let args: Array<string>;
	let repoRoot: string;

	try {
		repoRoot = validateRepoPath(data['repo_path']);
		args = buildArgs(String(data['action'] ?? ''), repoRoot, data);
	}
	catch (e: unknown) {
		ws.send(JSON.stringify({
			requestId,
			status: 'error',
			message: e instanceof Error ? e.message : 'Invalid rebase request',
		}));

		return;
	}

	const proc = Bun.spawn(['git', ...args], {
		cwd: repoRoot,
		env: getAgentEnv(ws),
		stdout: 'pipe',
		stderr: 'pipe',
	});

	const [stdout, stderr] = await Promise.all([
		new Response(proc.stdout).text(),
		new Response(proc.stderr).text(),
	]);
	const exitCode = await proc.exited;

	if (exitCode === 0) {
		ws.send(JSON.stringify({requestId, status: 'success', data: stdout}));

		return;
	}

	ws.send(JSON.stringify({
		requestId,
		status: 'error',
		message: stderr.trim() || `git rebase exited ${exitCode}`,
	}));
}

File: packages/remote-worker-rs/src/commands/git_rebase.rs
use crate::commands::resolve_file_path;
use crate::protocol;

fn build_args(
	action: &str,
	repo_path: &std::path::Path,
	req: &protocol::WsRequest,
) -> Result<Vec<String>, String> {
	if action == "continue" {
		return Ok(vec![
			"-c".into(),
			"core.editor=true".into(),
			"rebase".into(),
			"--continue".into(),
		]);
	}

	if action != "start" {
		return Err(format!("Unknown rebase action: {action}"));
	}

	let upstream = req
		.payload
		.get("upstream")
		.and_then(|v| v.as_str())
		.filter(|s| !s.is_empty())
		.ok_or("upstream must be a non-empty string")?;

	if upstream.starts_with('-') {
		return Err("upstream must not start with \"-\"".into());
	}

	let todo_raw = req
		.payload
		.get("todo_path")
		.and_then(|v| v.as_str())
		.filter(|s| !s.is_empty())
		.ok_or("todo_path must be a non-empty string")?;

	// The path is interpolated into a shell command via sequence.editor, so a single
	// quote would break out of the surrounding quotes.
	if todo_raw.contains('\'') {
		return Err("todo_path must not contain a single quote".into());
	}

	let todo = resolve_file_path(repo_path, todo_raw)?;

	if !todo.exists() {
		return Err(format!("Todo file does not exist: {}", todo.display()));
	}

	Ok(vec![
		"-c".into(),
		"core.editor=false".into(),
		"-c".into(),
		"rebase.missingCommitsCheck=ignore".into(),
		"-c".into(),
		format!("sequence.editor=cp '{}'", todo.display()),
		"rebase".into(),
		"-i".into(),
		"--autostash".into(),
		"--".into(),
		upstream.to_string(),
	])
}

pub async fn run(req: &protocol::WsRequest) -> String {
	let request_id = &req.request_id;

	let repo_path = match req.payload.get("repo_path").and_then(|v| v.as_str()) {
		Some(p) if !p.is_empty() => std::path::PathBuf::from(p),
		_ => return protocol::error(request_id, "repo_path must be a non-empty string"),
	};

	if !repo_path.exists() {
		return protocol::error(
			request_id,
			&format!("Repository path does not exist: {}", repo_path.display()),
		);
	}

	let action = req.payload.get("action").and_then(|v| v.as_str()).unwrap_or("");

	let args = match build_args(action, &repo_path, req) {
		Ok(a) => a,
		Err(e) => return protocol::error(request_id, &e),
	};

	let output = tokio::process::Command::new("git")
		.args(&args)
		.current_dir(&repo_path)
		.output()
		.await;

	match output {
		Ok(out) if out.status.success() => {
			protocol::success(request_id, serde_json::json!(String::from_utf8_lossy(&out.stdout)))
		}
		Ok(out) => {
			let stderr = String::from_utf8_lossy(&out.stderr).trim().to_string();
			protocol::error(
				request_id,
				if stderr.is_empty() { "git rebase failed" } else { &stderr },
			)
		}
		Err(e) => protocol::error(request_id, &format!("Failed to run git: {e}")),
	}
}
